ci: Run fuzzer task for the master branch only

hebasto opened the pull request:
https://github.com/bitcoin/bitcoin/pull/22629#issuecomment-896454270:
> I think we need to decide whether running the fuzzer CI in any branch other than master is something we want to be doing / maintaining. This seems pretty unsustainable unless we at least make changes in regards to the fuzz inputs being used by the different branches. I'm pretty sure Marco has mentioned this before.

This PR makes CI ignore fuzz tests by forcing `RUN_FUZZ_TESTS=false` for all cases when it is not the master branch or a PR based on it.

See #22731 as a demo for the 22.x branch.

hebasto commented:
cc @fanquake @MarcoFalke

jamesob commented:
ACK https://github.com/bitcoin/bitcoin/pull/22730/commits/cac87edab2acbccaa28d46ca0aa3b03ad770a6cc

Zero-1729 commented:
Concept ACK

> if  [ "${CIRRUS_BRANCH}" == "${CIRRUS_DEFAULT_BRANCH}" ] || [ "${CIRRUS_BASE_BRANCH}" == "${CIRRUS_DEFAULT_BRANCH}" ]; then

I think the second condition where it checks if the PR is based on the `master` base branch should also include an additional check to see if the PR affects the fuzzing system, to narrow it down (i.e. like checking the `CIRRUS_CHANGE_TITLE` to see if the PR's title is prepended with `fuzz:` for example, or contains the word `fuzz` elsewhere in the title). What do you think @hebasto?

fanquake commented:
> I think the second condition where it checks if the PR is based on the master branch should also include an additional check to see if the PR affects the fuzzing system, to narrow it down (i.e. like checking the CIRRUS_CHANGE_TITLE to see if the PR's title is prepended with fuzz: for example, or contains the word fuzz elsewhere in the title).

That sounds broken / over-complicated. On the master branch the fuzzers should be running on all commits, regardless of if the change is to the fuzzers themselves, or has a certain commit prefix. All we want to do is exclude branches other than master.

Zero-1729 commented:
Ah, thanks for the clarification! Seems I misunderstood the OP.

In that case, ignore my comment, the change looks good to go.

Zero-1729 reviewed (commented):
crACK cac87edab2acbccaa28d46ca0aa3b03ad770a6cc

MarcoFalke commented:
Ideally the fuzzers would run on all branches, it is just that my personal efforts are focussed on the master branch. While the fuzzers should run on older branches, I won't be working on this actively myself.

About the changes: The changes here won't disable the fuzzers when built locally or run in CI locally. If the goal is to disable the fuzz task in our Cirrus CI, it might be better to just do that in the cirrus.yml and not by putting CI-specific symbols (`CIRRUS_*`) in our CI-agnostic scripts (`./ci`).

hebasto commented:
Updated cac87edab2acbccaa28d46ca0aa3b03ad770a6cc -> fa2234fb70d4b889dd0c3ee4de9fa2afbec0af14 ([pr22730.01](https://github.com/hebasto/bitcoin/commits/pr22730.01) -> [pr22730.02](https://github.com/hebasto/bitcoin/commits/pr22730.02)).

Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/22730#issuecomment-900816207):
> If the goal is to disable the fuzz task in our Cirrus CI, it might be better to just do that in the cirrus.yml and not by putting CI-specific symbols (`CIRRUS_*`) in our CI-agnostic scripts (`./ci`).

hebasto committed:
ci: Run fuzzer task for the master branch only

laanwj commented:
Concept ACK, although in theory it'd make sense to fuzz some things across versions, I agree in practice it's unmaintainable, it trips up the CI for unrelated reasons too often.

MarcoFalke commented on .cirrus.yml:
```diff
@@ -42,7 +42,7 @@ global_task_template: &GLOBAL_TASK_TEMPLATE
     folder: "depends/built"
   ci_script:
-    - ./ci/test_run_all.sh
+    - if [ "${CIRRUS_BRANCH}" != "${CIRRUS_DEFAULT_BRANCH}" ] && [ "${CIRRUS_BASE_BRANCH}" != "${CIRRUS_DEFAULT_BRANCH}" ]; then export RUN_FUZZ_TESTS="false"; fi && source ./ci/test_run_all.sh
```
I meant that the whole task can be skipped. Compilation will be checked by the other tasks already and there is nothing else done, no?

hebasto replied:
Right. And this approach has one line diff :)

Done in the recent push.

MarcoFalke commented:
cr ACK 5a9e255e5a324e7aa0b63a9634aa3cfda9a300bd no opinion on the concept, also didn't test

hebasto commented:
> ... also didn't test

#22731 could be useful.

fanquake reviewed (approved):
ACK 5a9e255e5a324e7aa0b63a9634aa3cfda9a300bd - didn't test other than to look at #22731.

hebasto commented:
@fanquake 

Backport to 22.0 (and drop all other CI fuzz task fixes)?

fanquake commented:
> Backport to 22.0 (and drop all other CI fuzz task fixes)?

https://github.com/bitcoin/bitcoin/pull/22629#issuecomment-902489016

hebasto commented:
Backported in #22629.
